dataset/bin.cpp

```diff
@@ -351,21 +351,30 @@ auto axis_actions(const VariableConstView &data, const Coords &coords,
   return builder;
 }
 
-template <class Masks>
-auto hide_masked(const VariableConstView &data, const Masks &masks,
-                 const Dimensions &dims) {
-  const auto &[begin_end, buffer_dim, buffer] =
-      data.constituents<core::bin<DataArray>>();
-  auto [begin, end] = unzip(begin_end);
-  for (const auto dim : dims.labels()) {
-    auto mask = irreducible_mask(masks, dim);
-    if (mask) {
-      begin *= ~mask;
-      end *= ~mask;
+class HideMasked {
```

Comment:
The problem is that it returns a non-owning binned variable, and we need something to keep the indices alive. Previously this did not matter due to the "bug" of missing `is_view` specializations, but fixing this exposed the issue here.

Reply:
Hm, this is a similar issue to my comment here: https://github.com/scipp/scipp/pull/1488#discussion_r545862744
Maybe it would help to have a helper RAII class to create a temporary owner?

Reply:
Isn't `HideMasked` that RAII class? Or what exactly do you have in mind?

Reply:
I meant something more general. But that was just an idea I had in the moment. Not sure if there is actually a good way of implementing this.